### PR TITLE
[B] Account for spacing in MapFont#getWidth(). Fixes BUKKIT-4089

### PR DESCRIPTION
**Pull request by cnr**
(Note that [someone else did a PR for this before](https://github.com/Bukkit/Bukkit/pull/768), but that one doesn't follow the new PR conventions)
### The issue

The implementation of MapFont#getWidth(String) does not accurately reflect the width of the input, unless the input consists of 1 or less characters.

This is evident when looking at CraftBukkit's implementation of MapCanvas#drawText(), which contains [this line](https://github.com/Bukkit/CraftBukkit/blob/61e57e1196fc917d0e8f9c5f1ee48818a85fd5b9/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java#L106):

```
x += sprite.getWidth() + 1;
```

This line defines a 1-px space added after each character, which is not reflected in [the current MapFont#getWidth()](https://github.com/Bukkit/Bukkit/blob/dce83b6ce48c0c8592e7912e9527080254a22553/src/main/java/org/bukkit/map/MapFont.java#L48).
### PR Breakdown

This change first checks if the String contains 0 characters (as to not inaccurately report a width of -1) and returns `0` if this is the case.

After the character widths are added together, `text.length() - 1` is added to account for the 1-px spacing between characters
### Testing results and Materials

[The original PR submitter had a compelling test case](https://github.com/Bukkit/Bukkit/pull/768#issuecomment-13403025).

In addition, see [CraftMapCanvas#drawText()](https://github.com/Bukkit/CraftBukkit/blob/61e57e1196fc917d0e8f9c5f1ee48818a85fd5b9/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java#L106) to identify that there are no edge cases in regard to spacing the characters; _every character_ has a 1-px spacer appended to the end.
### JIRA Ticket

BUKKIT-4089 - https://bukkit.atlassian.net/browse/BUKKIT-4089

### CODE REVIEW

**Comment by riking** [comment]
To reiterate from the other PR, color codes are not permitted in the input to this method.

**Comment by AmoebaMan** [comment]
Why not permit color codes?  Why not intelligently filter them out?  It's not difficult to check if the preceding character is U+00A7.

**Comment by Black-Hole** [comment]
ChatColor.stripColor() should do the job.

**Comment by riking** [comment]
@AmoebaMan It already throws an exception when you give it a string with a color code in it - therefore we can assume no plugins are already calling the method with a color-code string. @Black-Hole Yes, either that or explicitly throwing an IllegalArgumentException.

**Comment by cnr** [comment]
@Black-Hole ChatColor#stripColor() would not work: map color codes [are in the format of "\u00A7[0-9]+;"](https://github.com/Bukkit/CraftBukkit/blob/master/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java#L86) (e.g. "§36;" for LIGHT_GRAY).

As @riking mentioned, plugins calling this already expect that they can't insert color codes (and given how fundamentally broken this method is, I doubt many plugins are using it at all). That said, I'll likely append a commit to add this functionality anyway (and add a much-needed built-in method to strip map color codes)

**Comment by AmoebaMan** [comment]
@riking Yes, which is a very silly behavior.  Why would you throw errors and force developers to strip color codes themselves when we could be doing it right here with almost no effort?  How is that helpful?

**Comment by riking** [comment]
Sure, go ahead and change it.

By the way, newlines don't have any width either. Although _that_ might qualify as exceptional.

**Comment by AmoebaMan** [comment]
Again, why throw exceptions when you could fairly simply add functionality to cope with it?  Split the string by newlines and return the length of the longest segment.

**Comment by cnr** [comment]
Appended a commit for newlines and color codes.

A couple of test cases (calculated programmatically):

---

`"Test§36;ing\nthis"`
- Color gets stripped -> `"Testing\nthis"`
- Lines are split -> `{"Testing", "this"}`
- `Testing` -> 35. ([which is indeed the correct result](https://github.com/Bukkit/Bukkit/pull/768#issuecomment-13403025))
- `this` -> 17.
- Result: 35.

---

`"t§3;h§66;is\nTe\nst§36;ing"`
- Color gets stripped -> `"this\nTe\nsting"`
- Lines are split -> `{"this", "Te", "sting"}`
- `this` -> 17
- `Te` -> 11
- `sting` -> 23
- Result: 23

**Comment by AmoebaMan** [comment]
Nice.  If it meets Bukkit's code specs, I think it ought to be pulled.

File: src/main/java/org/bukkit/map/MapPalette.java [hunk 1]
```diff
@@ -4,11 +4,14 @@
...
 import java.awt.Image;
 import java.awt.image.BufferedImage;
+import java.util.regex.Pattern;
 
 /**
  * Represents the palette that map items use.
  */
 public final class MapPalette {
+    public static final char COLOR_CHAR = '\u00A7';
+    private static final Pattern STRIP_COLOR_PATTERN = Pattern.compile(String.valueOf(COLOR_CHAR) + "[0-9]+;");
```
Comment (spaceemotion) [inline src/main/java/org/bukkit/map/MapPalette.java]: I thought ChatColor already has a pattern and a function to strip out all colors from a text?

Comment (cnr) [inline src/main/java/org/bukkit/map/MapPalette.java]: It does, but the color codes for text do not match those for map text. The pattern for map text color is "§[0-9]+;" instead of ChatColor's "§[0-9A-FK-OR]"

Comment (ST-DDT) [inline src/main/java/org/bukkit/map/MapPalette.java]: But i guess it still should be added to ChatColor.
Including the same for Format Codes (although not required)
Maybe split it to another PR including the strip methods stripColorCodesColors and stripColorCodesFormats

centralise logic/code to simplify maintenance

Comment (cnr) [inline src/main/java/org/bukkit/map/MapPalette.java]: While I agree that redundant code isn't optimal, I don't believe that putting this functionality in ChatColor is the proper solution.

Separating the color stripping method from MapPalette wouldn't be an appropriate course of action unless we were to also migrate the color codes, which most certainly do not belong in ChatColor.

If anything, we'd have to move such a function to a utility class and call it from both ChatColor and MapPalette.

Comment (ST-DDT) [inline src/main/java/org/bukkit/map/MapPalette.java]: Your are right i only had a quick view when writing my post.
Now i spend more time reviewing your commit and placing your code there is definitly a good solution

Comment (EdGruberman) [inline src/main/java/org/bukkit/map/MapPalette.java]: I'm still unclear on why ChatColor.stripColor is insufficient for your purposes here?  A § symbol in Minecraft code refers to a chat formatting character, be it color or style.  Under this context, treating that character similarly, but with different filtering characteristics diverges it's meaning from the centralized ChatColor implementation.  What is the justification for that?  Is there some use case in the Map API that needs to treat §A as two individual characters and not a formatting code? 

After my quick review of the Map API it doesn't even seem like chat formatting codes are properly treated anywhere in it.  Perhaps I'm overlooking something?  But even if it is, it should be treated like a chat formatting code is elsewhere and not separately unless a use case exists I'm not clear on.

Does the Map API even color text according to chat formatting codes anywhere?  If not, the § character should actually generate an exception itself as not a valid character and the `isValid` method should be updated accordingly.

Comment (riking) [inline src/main/java/org/bukkit/map/MapPalette.java]: Maps use 1 of ~~15~~14 colors, selected from https://github.com/eueln/Bukkit/blob/7040d36d66f4b4e3b5de2556489cceb03f789bb6/src/main/java/org/bukkit/map/MapPalette.java#L34

Comment (EdGruberman) [inline src/main/java/org/bukkit/map/MapPalette.java]: What code translates chat color codes into those array indexes?

Comment (cnr) [inline src/main/java/org/bukkit/map/MapPalette.java]: ChatColor#stripColor is insufficient because _map color codes and chat color codes are fundamentally different_. To demonstrate this, an example chat color code is `§a`, and an example map color code is `§36;` which _would not_ get stripped by ChatColor#stripColor

The only other place in Bukkit-land that I saw \u00A7 mentioned was MapFont#isValid(String), otherwise in the reference implementation (CraftBukkit) [it is established that the proper way to format color codes is §[0-9]+;](https://github.com/Bukkit/CraftBukkit/blob/master/src/main/java/org/bukkit/craftbukkit/map/CraftMapCanvas.java#L86).

I believe it is out of the realm of this PR, but there is some restructuring that needs to be done for MapPalette to give a ChatColor-like way to insert color codes in text and to better detail the way that color codes work. This was just meant to be a quick PR to fix the fundamentally broken MapFont#getWidth()

Comment (jrtc27) [inline src/main/java/org/bukkit/map/MapPalette.java]: Should this not be using `Pattern.quote`, as `COLOR_CHAR` could theoretically be a regex special character (even though it isn't here)?

Comment (cnr) [inline src/main/java/org/bukkit/map/MapPalette.java]: @jrtc27 I just did it [the same way that ChatColor does it](https://github.com/Bukkit/Bukkit/blob/master/src/main/java/org/bukkit/ChatColor.java#L108)

Comment (jrtc27) [inline src/main/java/org/bukkit/map/MapPalette.java]: Fair enough. It's not necessary, I just thought it might be better practice.

Comment (EdGruberman) [inline src/main/java/org/bukkit/map/MapPalette.java]: Thank you for clarifying @eueln.

File: src/main/java/org/bukkit/map/MapFont.java [hunk 2]
```diff
@@ -50,10 +50,28 @@ public int getWidth(String text) {
...
+        }
+
         int result = 0;
-        for (int i = 0; i < text.length(); ++i) {
-            result += chars.get(text.charAt(i)).getWidth();
+
+        String[] lines = text.split("\n");
+        for (String line : lines) {
+            int length = 0;
+
```
Comment (riking) [inline src/main/java/org/bukkit/map/MapFont.java]: You have spaces on this line.

**Comment by turt2live** [comment]
@eueln If you are still interested in this please update it to the latest code base. As well the CONTRIBUTING.md guidelines will need to be followed.

Once these concerns are addressed the review process may continue.

**Comment by cnr** [comment]
@turt2live Rebased. This PR ended up branching into two issues, which is what the two commits represent. Should I squash them further into one more general commit, or is this good as-is? Is there anything else that needs to be done?

**Comment by turt2live** [comment]
I'm not too sure if the second (colors and new lines) change should be included - I'll get back to you on this.

As it stands right now, I just need to do some internal work and we'll be rolling :)

File: src/main/java/org/bukkit/map/MapFont.java [hunk 3]
```diff
@@ -50,10 +50,28 @@ public int getWidth(String text) {
...
+            for (int i = 0; i < line.length(); ++i) {
+                length += chars.get(line.charAt(i)).getWidth();
+            }
+            length += line.length() - 1; // Account for 1-px spaces between characters
+
+            if (length > result) {
+                result = length;
+            }
         }
+
```
Comment (turt2live) [inline src/main/java/org/bukkit/map/MapFont.java:64]: This line will have to be removed

**Comment by turt2live** [comment]
We would like these 2 fixes be separated into 2 different PRs please.

**Comment by cnr** [comment]
Given that the second commit/would-be-PR is reliant on code from the first, should I remove the second commit, wait for this to be pulled, then submit the second PR?

(whoops, didn't mean to close the PR)

**Comment by turt2live** [comment]
Remove the second commit from here.
Open a new PR for it
Make it all pretty and up to standard
Mention it requires this PR
Ping me in it.

**Comment by turt2live** [comment]
This PR still contains the second commit for formatting. Has this been split up? If so, please link the second PR here.

Otherwise this looks pretty much good to go.

**Comment by cnr** [comment]
Updated with the second commit removed. If/when this is pulled, I'll open another PR for the second commit.

**Comment by turt2live** [comment]
Pulled, thanks!

https://github.com/Bukkit/Bukkit/commit/236ebabda2a4be112181a1c3dd537c5919af3d41

**Comment by turt2live** [comment]
Apparently this PR was never closed even though it was actually pulled. I'll just close it now... 

Thanks!
